feat: add C ndarray API and refactor `blas/ext/base/dsortins`

0PrashantYadav0 opened the pull request:
Resolves None.

## Description

> What is the purpose of this pull request?

This pull request:

-   add C ndarray API and refactor `blas/ext/base/dsortins`

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves None

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

0PrashantYadav0 committed:
feat: add C ndarray interface and refactor implementation

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
---

0PrashantYadav0 commented:
/stdlib update-copyright-years

0PrashantYadav0 commented:
/stdlib lint-autofix

stdlib-bot committed:
fix: resolve lint errors

0PrashantYadav0 committed:
fix: correct argument order in dsortins function and example

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: failed
---

0PrashantYadav0 commented:
/stdlib merge

stdlib-bot committed:
Merge remote-tracking branch 'upstream/develop' into refactor-dsortins

0PrashantYadav0 committed:
chore: updated copyright year

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: passed
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: passed
  - task: run_c_examples
    status: failed
---

0PrashantYadav0 committed:
fix: remove return statement in stdlib_strided_dsortins function

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: passed
  - task: run_c_examples
    status: failed
---

0PrashantYadav0 commented:
/stdlib update-copyright-years

0PrashantYadav0 commented:
@kgryte @headlessNode @aayush0325 will you please help me to resolve the error in github cli.

gunjjoshi commented on lib/node_modules/@stdlib/blas/ext/base/dsortins/benchmark/c/benchmark.unsorted_random.length.c:
@0PrashantYadav0 Looks like there is a copy-paste error in this file, you've added the whole code twice in this benchmark. This is what is leading to the linting errors.

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/dsortins/README.md:
````diff
@@ -153,6 +153,118 @@ console.log( x );
+```c
+double x[] = { 1.0, -2.0, 3.0, -4.0 };
+
+stdlib_strided_dsortins( 2, -1, x, 1 );
````
```suggestion
stdlib_strided_dsortins( 2, -1.0, x, 1 );
```

The input type is a `double`. In which case, include the decimal.

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/dsortins/README.md:
````diff
@@ -153,6 +153,118 @@ console.log( x );
+```c
+double x[] = { 1.0, -2.0, 3.0, -4.0 };
+
+stdlib_strided_dsortins_ndarray( 4, 1, x, 1, 0 );
````
```suggestion
stdlib_strided_dsortins_ndarray( 4, 1.0, x, 1, 0 );
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/dsortins/README.md:
```diff
@@ -153,6 +153,118 @@ console.log( x );
+The function accepts the following arguments:
+
+-   **N**: `[in] CBLAS_INT` number of indexed elements.
+-   **order**: `[in] CBLAS_INT` sort order.
```
```suggestion
-   **order**: `[in] double` sort order. If `order < 0.0`, the input strided array `x` is sorted in **decreasing** order. If `order > 0.0`, the input strided array `x` is sorted in **increasing** order. If `order == 0.0`, the input strided arrays are left unchanged.
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/dsortins/README.md:
```diff
@@ -153,6 +153,118 @@ console.log( x );
+
+-   **N**: `[in] CBLAS_INT` number of indexed elements.
+-   **order**: `[in] CBLAS_INT` sort order.
+-   **X**: `[inout] double*` input array. If `order < 0.0`, the input strided array `x` is sorted in **decreasing** order. If `order > 0.0`, the input strided array `x` is sorted in **increasing** order. If `order == 0.0`, the input strided arrays are left unchanged.
```
```suggestion
-   **X**: `[inout] double*` input array.
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/dsortins/benchmark/c/benchmark.unsorted_random.length.c:
```diff
@@ -0,0 +1,381 @@
+	}
+	t = tic();
+	for ( i = 0; i < iterations; i++ ) {
+		stdlib_strided_dsortins( len, 1, x, 1 );
```
```suggestion
		stdlib_strided_dsortins( len, 1.0, x, 1 );
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/dsortins/benchmark/c/benchmark.unsorted_random.length.c:
```diff
@@ -0,0 +1,381 @@
+	}
+	t = tic();
+	for ( i = 0; i < iterations; i++ ) {
+		stdlib_strided_dsortins_ndarray( len, 1, x, 1, 0);
```
```suggestion
		stdlib_strided_dsortins_ndarray( len, 1.0, x, 1, 0);
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/dsortins/benchmark/c/benchmark.unsorted_random.length.c:
```diff
@@ -0,0 +1,381 @@
+		}
+	}
+	print_summary( count, count );
+}/**
```
@0PrashantYadav0 Remove the duplicate code below.

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/dsortins/docs/repl.txt:
```diff
@@ -1,8 +1,8 @@
     Sorts a double-precision floating-point strided array using insertion sort.
 
-    The `N` and `stride` parameters determine which elements in `x` are accessed
+    The `N` and `stride` parameters determine which elements in are accessed
```
```suggestion
    The `N` and stride parameters determine which elements in are accessed
```

This and Line 6 need to be rewrapped at 80 chars.

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/dsortins/docs/repl.txt:
```diff
@@ -61,25 +61,24 @@
 
     While typed array views mandate a view offset based on the underlying
-    buffer, the `offset` parameter supports indexing semantics based on a
+    buffer, the `offsetX` parameter supports indexing semantics based on a
```
```suggestion
    buffer, the offset parameter supports indexing semantics based on a
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/dsortins/include/stdlib/blas/ext/base/dsortins.h:
```diff
@@ -31,7 +31,12 @@ extern "C" {
 * Sorts a double-precision floating-point strided array using insertion sort.
 */
-void c_dsortins( const int64_t N, const double order, double *X, const int64_t stride );
+void API_SUFFIX(stdlib_strided_dsortins)( const CBLAS_INT N, const double order, double *X, CBLAS_INT strideX );
```
```suggestion
void API_SUFFIX(stdlib_strided_dsortins)( const CBLAS_INT N, const double order, double *X, const CBLAS_INT strideX );
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/dsortins/include/stdlib/blas/ext/base/dsortins.h:
```diff
@@ -31,7 +31,12 @@ extern "C" {
+/**
+* Sorts a double-precision floating-point strided array using insertion sort and alternative indexing semantics.
+*/
+void API_SUFFIX(stdlib_strided_dsortins_ndarray)( const CBLAS_INT N, const double order, double *X, CBLAS_INT strideX, CBLAS_INT offsetX );
```
```suggestion
void API_SUFFIX(stdlib_strided_dsortins_ndarray)( const CBLAS_INT N, const double order, double *X, const CBLAS_INT strideX, const CBLAS_INT offsetX );
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/dsortins/lib/dsortins.js:
```diff
@@ -43,101 +43,9 @@ var isnan = require( '@stdlib/math/base/assert/is-nan' );
-	return x;
+function dsortins( N, order, x, strideX ) {
+	var offsetX = stride2offset( N, strideX );
+	return ndarray( N, order, x, strideX, offsetX );
```
```suggestion
	return ndarray( N, order, x, strideX, stride2offset( N, strideX ) );
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/dsortins/lib/ndarray.native.js:
```diff
@@ -43,15 +42,8 @@ var addon = require( './dsortins.native.js' );
-	view = offsetView( x, offset );
-	addon( N, order, view, stride );
+function dsortins( N, order, x, strideX, offsetX ) {
+	addon.ndarray( N, order, x, strideX, offsetX);
```
```suggestion
	addon.ndarray( N, order, x, strideX, offsetX );
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/dsortins/src/main.c:
```diff
@@ -28,15 +29,28 @@
-	int64_t lx;
-	int64_t sx;
-	int64_t i;
+void API_SUFFIX(stdlib_strided_dsortins)( const CBLAS_INT N, const double order, double *X, CBLAS_INT strideX) {
```
```suggestion
void API_SUFFIX(stdlib_strided_dsortins)( const CBLAS_INT N, const double order, double *X, const CBLAS_INT strideX ) {
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/dsortins/src/main.c:
```diff
@@ -28,15 +29,28 @@
+* @param strideX  index increment
+* @param offsetX  starting index
+*/
+void API_SUFFIX(stdlib_strided_dsortins_ndarray)( const CBLAS_INT N, const double order, double *X, CBLAS_INT strideX, CBLAS_INT offsetX ) {
```
```suggestion
void API_SUFFIX(stdlib_strided_dsortins_ndarray)( const CBLAS_INT N, const double order, double *X, const CBLAS_INT strideX, const CBLAS_INT offsetX ) {
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/dsortins/src/main.c:
```diff
@@ -46,15 +60,15 @@ void c_dsortins( const int64_t N, const double order, double *X, const int64_t s
-		sx = -stride;
-	} else {
-		sx = stride;
+		strideX *= -1;
```
You need to use `sx`, as in the original. There is no reason for the input `strideX` argument to not be a constant.

0PrashantYadav0 committed:
fix: correct argument types and improve parameter descriptions in dsortins functions

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: passed
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: passed
  - task: run_c_benchmarks
    status: failed
---

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/dsortins/docs/repl.txt:
```diff
@@ -1,8 +1,8 @@
 
-    The `N` and `stride` parameters determine which elements in `x` are accessed
+    The `N` and stride parameters determine which elements in are accessed
     at runtime.
```
```suggestion
    The `N` and stride parameters determine which elements in the strided array
    are accessed at runtime.
```

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/dsortins/docs/repl.txt:
```diff
@@ -94,10 +93,10 @@
 
-    offset: integer
+    offsetX: integer
         Starting index of `x`.
```
```suggestion
        Starting index.
```

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/dsortins/README.md:
```diff
@@ -48,9 +48,9 @@ The function has the following parameters:
+-   **strideX**: stride length.
 
-The `N` and `stride` parameters determine which elements in `x` are accessed at runtime. For example, to sort every other element
+The `N` and stride parameters determine which elements in `x` are accessed at runtime. For example, to sort every other element
```
```suggestion
The `N` and stride parameters determine which elements in the strided array are accessed at runtime. For example, to sort every other element:
```

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/dsortins/README.md:
````diff
@@ -153,6 +153,118 @@ console.log( x );
+-   **strideX**: `[in] CBLAS_INT` stride length for `X`.
+
+```c
+stdlib_strided_dsortins( const CBLAS_INT N, const double order, double *X, CBLAS_INT strideX );
````
```suggestion
stdlib_strided_dsortins( const CBLAS_INT N, const double order, double *X, const CBLAS_INT strideX );
```

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/dsortins/README.md:
````diff
@@ -153,6 +153,118 @@ console.log( x );
+-   **offsetX**: `[in] CBLAS_INT` starting index for `X`.
+
+```c
+stdlib_strided_dsortins_ndarray( const CBLAS_INT N, const double order, double *X, CBLAS_INT strideX, CBLAS_INT offsetX );
````
```suggestion
stdlib_strided_dsortins_ndarray( const CBLAS_INT N, const double order, double *X, const CBLAS_INT strideX, const CBLAS_INT offsetX );
```

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/dsortins/benchmark/c/benchmark.unsorted_random.length.c:
```diff
@@ -0,0 +1,191 @@
+	}
+	t = tic();
+	for ( i = 0; i < iterations; i++ ) {
+		stdlib_strided_dsortins( len, 1, x, 1 );
```
```suggestion
		stdlib_strided_dsortins( len, 1.0, x, 1 );
```

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/dsortins/src/main.c:
```diff
@@ -28,15 +29,30 @@
-	int64_t lx;
-	int64_t sx;
-	int64_t i;
+void API_SUFFIX(stdlib_strided_dsortins)( const CBLAS_INT N, const double order, double *X, CBLAS_INT strideX) {
```
```suggestion
void API_SUFFIX(stdlib_strided_dsortins)( const CBLAS_INT N, const double order, double *X, const CBLAS_INT strideX ) {
```

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/dsortins/src/main.c:
```diff
@@ -28,15 +29,30 @@
+* @param strideX  index increment
+* @param offsetX  starting index
+*/
+void API_SUFFIX(stdlib_strided_dsortins_ndarray)( const CBLAS_INT N, const double order, double *X, CBLAS_INT strideX, CBLAS_INT offsetX ) {
```
```suggestion
void API_SUFFIX(stdlib_strided_dsortins_ndarray)( const CBLAS_INT N, const double order, double *X, const CBLAS_INT strideX, const CBLAS_INT offsetX ) {
```

headlessNode commented:
@0PrashantYadav0 I just noticed, the JS examples. You will need to update the JS example in the `README.md` and `examples/index.js`

0PrashantYadav0 committed:
chore: updated files according to code review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: passed
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: passed
  - task: run_c_benchmarks
    status: failed
---

0PrashantYadav0 commented:
@headlessNode Thank you for your review. I have updated everything. Please review and merge this pr.

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/dsortins/README.md:
```diff
@@ -153,6 +154,118 @@ console.log( x );
+    int strideX = 1;
+
+    // Sort the array:
+    API_SUFFIX(stdlib_strided_dsortins)( N, 1.0, x, strideX );
```
```suggestion
    stdlib_strided_dsortins( N, 1.0, x, strideX );
```

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/dsortins/benchmark/c/benchmark.unsorted_random.length.c:
```diff
@@ -0,0 +1,191 @@
+	}
+	t = tic();
+	for ( i = 0; i < iterations; i++ ) {
+		stdlib_strided_dsortins_ndarray( len, 1.0, x, 1, 0);
```
```suggestion
		stdlib_strided_dsortins_ndarray( len, 1.0, x, 1, 0 );
```

headlessNode reviewed (changes requested):
Apart these two suggestions, LGTM.

0PrashantYadav0 committed:
fix: remove API_SUFFIX macro in dsortins examples and benchmarks for clarity

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: passed
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: passed
  - task: run_c_benchmarks
    status: failed
---

Planeshifter reviewed (approved):
Thanks @0PrashantYadav0 for your contribution and @kgryte & @headlessNode for reviews!
